### PR TITLE
Adds per-user queue length limits

### PR DESCRIPTION
**Pull request by dposada**
## Changes proposed in this PR

- adding per-user queue length limits
- pools can be configured with a "normal" per-user limit, and a "constrained" per-user limit, where the constrained number is used when the entire pool's queue length is above a configurable threshold
- for example, you could set pool `k8s-alpha` to allow each user to have 50k jobs queued if the entire `k8s-alpha` queue is <= 300k, but only allow each user to have 10k queued if the entire `k8s-alpha` queue is > 300k

## Why are we making these changes?

To prevent huge queues from impacting scheduling performance.

### CODE REVIEW

File: scheduler/src/cook/util.clj [hunk 1]
```diff
@@ -24,7 +24,8 @@
             [clojure.tools.namespace.dependency :refer :all]
             [postal.core :as postal]
             [schema.core :as s])
-  (:import (java.util.concurrent.atomic AtomicLong)))
+  (:import (java.util.concurrent.atomic AtomicLong)
```
Comment (scrosby) [inline scheduler/src/cook/util.clj:27]: Why AtomicLong?

File: scheduler/src/cook/queue_limit.clj [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            [cook.datomic :as datomic]
+            [cook.queries :as queries]
+            [cook.regexp-tools :as regexp-tools]
+            [cook.util :as util]
+            [datomic.api :as d]
+            [plumbing.core :as pc]
+            [metrics.timers :as timers]))
+
+(defn- match-based-on-pool-name
+  "Given a pool name and field name, returns
```
Comment (scrosby) [inline scheduler/src/cook/queue_limit.clj:15]: Maybe inline this into the caller or rename it? This function having the same name as regexp-tools/match-based-on-pool-name but behaving differently is awkward. 

My vote would be to make match-based-on-pool-name accept a default if not found, and then inline this into its callers, with maybe a helper function 

(def per-pool-config []
  (:per-pool (config/queue-limits)))

Then they become, e.g., :
   (match-based-on-pool-name (per-pool-config) pool-name :pool-global-threshold Integer/MAX_VALUE)

File: scheduler/src/cook/queue_limit.clj [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                              pool-b {user-x 20 user-y 40 user-z 140}
+                              ...}"
+  []
+  (let [pending-jobs
+        (get-pending-jobs)
+        pool->pending-jobs
+        (group-by
+          caches/job->pool-name
+          pending-jobs)]
+    {:pool->queue-length
```
Comment (scrosby) [inline scheduler/src/cook/queue_limit.clj:79]: Small optimization: If you do the per pool per-user map first, then you can reduce that to compute the overall length per pool.

File: scheduler/src/cook/queue_limit.clj [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+          pool-global-threshold (pool-global-threshold pool-name)]
+      (if (<= pool-global-length pool-global-threshold)
+        (user-limit-normal pool-name)
+        (user-limit-constrained pool-name))))
+
+  (defn inc-queue-length!
+    "Increments the pool-global and per-user queue lengths for
+    the given pool name and user by the given number of jobs"
+    [pool-name user number-jobs]
+    (let [inc-number-jobs #(-> % (or 0) (+ number-jobs))]
```
Comment (scrosby) [inline scheduler/src/cook/queue_limit.clj:117]: Probably worth an assert that pool-name is not nil.

File: scheduler/src/cook/rest/api.clj [hunk 5]
```diff
@@ -1969,6 +1973,50 @@
...
+  "Check if the job submission would cause the user to have more jobs queued than they're
+  allowed. We refresh queue lengths from the database on a configurable interval. Note that
+  this check does not take into account any of the following in real-time, and therefore is
+  only 100% accurate as of the last refresh from the database:
+
+  - Jobs will get removed from the queue when they launch
+  - Jobs will get added to the queue if they fail and have retries
+  - If there are multiple API hosts in the cluster, each one is maintaining a separate
+    count of the queue, without knowing about the other API hosts' job submissions
+
```
Comment (scrosby) [inline scheduler/src/cook/rest/api.clj:1986]: A suggestion to improve this is to give explicit error bars in the explanation.
"""
There are some errors in this:
   -- Jobs will get removed from teh queue then they launch. However, We limit users to 3-20 jobs/sec, so in 2 minutes a user can launch at most 200-2k jobs. 
   -- Jobs will get added to the queue if they are running, fail and have retries. Again, this is limited based on how many jobs a user has running.
   --- If there are multiple API hosts... This only lets a user go over-queue.
   --- If a user kills their jobs in the queue, they may be unable to reuse that for 2 minutes. To mitigiate this ....

However, we intend per-user queue limits to be in the 10k-50k range, so the errors are fairly minor, except for the kill jobs case.
"""

(We need to both describe and handle the kill jobs case IMO. My suggestion is hope for API server locality and decrement the queue on kill for a job that's still alive.)

File: scheduler/src/cook/util.clj [hunk 6]
```diff
@@ -190,3 +191,9 @@
   "Return triple of keys from two maps: [only in left, only in right, in both]"
   [left right]
   (data/diff (set (keys left)) (set (keys right))))
+
+(defn time-seq
```
Comment (scrosby) [inline scheduler/src/cook/util.clj:195]: Split this refactor into a separate PR?

File: scheduler/test/cook/test/tools.clj [hunk 7]
```diff
@@ -20,9 +20,11 @@
             [clojure.core.cache :as cache]
             [clojure.test :refer :all]
             [cook.config :as config]
+            [cook.queries :as queries]
             [cook.test.testutil :as testutil
              :refer [create-dummy-group create-dummy-instance create-dummy-job create-dummy-job-with-instances create-pool restore-fresh-database!]]
-            [cook.tools :as util]
+            [cook.tools :as tools]
```
Comment (scrosby) [inline scheduler/test/cook/test/tools.clj:26]: I feel the refactor of util to tools can also be split off.

File: scheduler/src/cook/regexp_tools.clj [hunk 8]
```diff
@@ -30,3 +30,9 @@
         (get field-name))
     (catch Exception e
       (throw (ex-info "Failed matching key" {:regexp-name regexp-name :field-name field-name :match-list match-list :key key} e)))))
+
+(defn match-based-on-pool-name
```
Comment (scrosby) [inline scheduler/src/cook/regexp_tools.clj:34]: This move could also be refactored into its own PR.

File: scheduler/src/cook/caches.clj [hunk 9]
```diff
@@ -28,4 +28,19 @@
 (mount/defstate ^Cache task-ent->user-cache :start (new-cache config/config))
 (mount/defstate ^Cache job-ent->user-cache :start (new-cache config/config))
 (mount/defstate ^Cache task->feature-vector-cache :start (new-cache config/config))
-(mount/defstate ^Cache job-uuid->dataset-maps-cache :start (new-cache config/config))
\ No newline at end of file
+(mount/defstate ^Cache job-uuid->dataset-maps-cache :start (new-cache config/config))
+
+(let [miss-fn
```
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/src/cook/caches.clj:34]: Good idea, will do

File: scheduler/src/cook/caches.clj [hunk 10]
```diff
@@ -28,4 +28,19 @@
...
+      (fn [{:keys [job/pool]}]
+        (or (:pool/name pool)
+            (config/default-pool)
+            "no-pool"))]
+  (defn job->pool-name
+    "Return the pool name of the job."
+    [job]
+    (lookup-cache-datomic-entity! job-ent->pool-cache miss-fn job)))
+
+(defn job-ent->user
```
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/src/cook/caches.clj:44]: :+1: will move

File: integration/tests/cook/test_multi_user.py [hunk 11]
```diff
@@ -937,3 +937,37 @@ def test_compute_cluster_api_delete_invalid(self):
...
+        bad_constraint = [["HOSTNAME", "EQUALS", "lol won't get scheduled"]]
+        user = os.getenv('COOK_TEST_QUEUE_LIMITS_USER')
+        user = self.user_factory.specific_user(user) if user else self.user_factory.new_user()
+        with user:
+            try:
+                util.kill_running_and_waiting_jobs(self.cook_url, user.name)
+                _, resp = util.submit_job(self.cook_url, pool=pool, constraints=bad_constraint)
+                self.assertEqual(resp.status_code, 201, resp.text)
+
+                _, resp = util.submit_jobs(self.cook_url, {'constraints': bad_constraint}, clones=max_limit, pool=pool)
```
[the remark this answers is not in the document]
Comment (dposada) [reply integration/tests/cook/test_multi_user.py:965]: Good idea, will do

File: scheduler/config-k8s.edn [hunk 12]
```diff
@@ -133,6 +133,11 @@
        :keystore-path #config/env "COOK_KEYSTORE_PATH"
        :keystore-type "pkcs12"
        :keystore-pass "cookstore"}
+ :queue-limits {:per-pool [{:pool-global-threshold 300000
+                            :pool-regex ".*"
```
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/config-k8s.edn:137]: Will do

File: integration/tests/cook/test_multi_user.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (dposada) [reply integration/tests/cook/test_multi_user.py]: Done

File: scheduler/config-k8s.edn [hunk 13]
```diff
@@ -133,6 +133,11 @@
        :keystore-path #config/env "COOK_KEYSTORE_PATH"
        :keystore-type "pkcs12"
        :keystore-pass "cookstore"}
+ :queue-limits {:per-pool [{:pool-global-threshold 300000
```
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/config-k8s.edn]: Will do

File: scheduler/src/cook/caches.clj (hunk 9 above)
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/src/cook/caches.clj]: Done

File: scheduler/src/cook/caches.clj (hunk 10 above)
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/src/cook/caches.clj]: Done

File: scheduler/src/cook/queries.clj [hunk 14]
```diff
@@ -28,3 +30,34 @@
             [?e :resource.type/mesos-name ?ident]]
           db)
        (map first)))
+
+(defn get-pending-job-ents*
```
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/src/cook/queries.clj:34]: Discussed offline the pros/cons of splitting this up, I decided to keep it as one PR

File: scheduler/src/cook/util.clj (hunk 6 above)
Comment (dposada) [reply scheduler/src/cook/util.clj:195]: Discussed offline the pros/cons of splitting this up, I decided to keep it as one PR

File: scheduler/test/cook/test/tools.clj (hunk 7 above)
Comment (dposada) [reply scheduler/test/cook/test/tools.clj:26]: Discussed offline the pros/cons of splitting this up, I decided to keep it as one PR

File: scheduler/src/cook/regexp_tools.clj (hunk 8 above)
Comment (dposada) [reply scheduler/src/cook/regexp_tools.clj:34]: Discussed offline the pros/cons of splitting this up, I decided to keep it as one PR

File: scheduler/src/cook/queue_limit.clj (hunk 2 above)
Comment (dposada) [reply scheduler/src/cook/queue_limit.clj:15]: Sounds good, will do

File: scheduler/src/cook/util.clj (hunk 1 above)
Comment (dposada) [reply scheduler/src/cook/util.clj:27]: No idea; the `AtomicLong` was already there, this is unchanged

File: scheduler/config-k8s.edn (hunk 12 above)
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/config-k8s.edn]: Done

File: scheduler/config-k8s.edn (hunk 13 above)
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/config-k8s.edn]: Done

File: scheduler/src/cook/queue_limit.clj (hunk 2 above)
Comment (dposada) [reply scheduler/src/cook/queue_limit.clj]: Done

File: scheduler/src/cook/queue_limit.clj (hunk 3 above)
Comment (dposada) [reply scheduler/src/cook/queue_limit.clj:86]: Done

File: scheduler/src/cook/queue_limit.clj (hunk 4 above)
Comment (dposada) [reply scheduler/src/cook/queue_limit.clj:121]: Done

File: scheduler/src/cook/rest/api.clj (hunk 5 above)
Comment (dposada) [reply scheduler/src/cook/rest/api.clj:1987]: Good call on killing jobs, I've now added that.

The error bars are dependent on the configuration you run Cook with, so I don't think we should give specific numbers about them in the code.

File: scheduler/test/cook/test/queue_limit.clj [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    (is (thrown? AssertionError (queue-limit/inc-queue-length! nil "alice" 2)))))
+
+(deftest test-dec-queue-length!
+  (with-redefs [queue-limit/get-pending-jobs (constantly [])]
+    (queue-limit/update-queue-lengths!))
+  (queue-limit/inc-queue-length! "pool-1" "alice" 2)
+  (queue-limit/inc-queue-length! "pool-1" "sally" 1)
+  (queue-limit/inc-queue-length! "pool-2" "alice" 1)
+  (queue-limit/inc-queue-length! "pool-2" "sally" 2)
+  (is (= 2 (queue-limit/user-queue-length "pool-1" "alice")))
```
Comment (scrosby) [inline scheduler/test/cook/test/queue_limit.clj:31]: I'd consider also putting a direct atom query here:

(is {:pool->queue-length {"pool-1" {"alice" 2} ....))  and ditto for other sections below.

Make sure the dictionary is what it should be. (Also by indicating the exact state stored, it makes it easier to check that the query functions following are returning the right result on that state.. Not to check them, I need to play all of the prior actions in my mind.

File: scheduler/src/cook/regexp_tools.clj [hunk 16]
```diff
@@ -30,3 +30,16 @@
...
+  "Given a list of dictionaries [{:pool-regexp .. :field ...} {:pool-regexp .. :field ...}
+   a pool name and a <field> name, return the first matching <field> where the regexp matches the pool name."
+  [match-list effective-pool-name field & {:keys [default-value] :or {default-value nil}}]
+  (let [value (match-based-on-regexp
```
Comment (scrosby) [inline scheduler/src/cook/regexp_tools.clj:38]: The defaulting won't work. When nothing matches, the regexp code above throws, that distinguishes between nil result and not specified. You could pass the default up to match-based-on-regexp and have it return that.

File: scheduler/test/cook/test/regexp_tools.clj [hunk 17]
```diff
@@ -24,3 +24,12 @@
...
 
+(deftest test-match-based-on-pool-name
+  (let [matchlist
+        [{:pool-regex "^foo$" :field {:foo 1}}
+         {:pool-regex ".*" :field {:bar 2}}
+         {:pool-regex "^baz$" :field {:baz 3}}]]
+    (is (= (regexp-tools/match-based-on-pool-name matchlist "foo" :field) {:foo 1}))
+    (is (= (regexp-tools/match-based-on-pool-name matchlist "bar" :field) {:bar 2}))
+    (is (= (regexp-tools/match-based-on-pool-name matchlist "baz" :field) {:bar 2})))
+  (is (= (regexp-tools/match-based-on-pool-name [] "foo" :field) nil)))
```
Comment (scrosby) [inline scheduler/test/cook/test/regexp_tools.clj:35]: Test default value code here please. 

There are a few interesting corner cases worth checking. E.g., a match that returns nil/false/{}/() vs no match, should return nil/false/{}/() and default.

File: scheduler/src/cook/rest/api.clj (hunk 5 above)
Comment (dposada) [reply scheduler/src/cook/rest/api.clj:1989]: Good call, done

File: integration/tests/cook/test_multi_user.py [hunk 18]
```diff
@@ -937,3 +939,66 @@ def test_compute_cluster_api_delete_invalid(self):
...
+            self.logger.info(f'Matching per-pool queue-limits in {pool} pool: {matching_configs}')
+
+        config = matching_configs[0]
+        max_limit = 5000
+        user_limit_normal = config['user-limit-normal']
+        user_limit_constrained = config['user-limit-constrained']
+        if (user_limit_normal > max_limit) or (user_limit_constrained > max_limit):
+            self.skipTest(f'Requires per-pool queue-limits in {pool} pool to be <= {max_limit}')
+
+        buffer = 10
```
[the remark this answers is not in the document]
Comment (dposada) [reply integration/tests/cook/test_multi_user.py]: Yep, renamed variable and added a comment, let me know if it's helpful

File: integration/tests/cook/test_multi_user.py [hunk 19]
```diff
@@ -937,3 +939,66 @@ def test_compute_cluster_api_delete_invalid(self):
...
+
+        def submission_succeeded(submit_resp):
+            self.assertEqual(submit_resp.status_code, 201, submit_resp.text)
+            return True
+
+        with user:
+            try:
+                # Kill running and waiting jobs and make sure we can
+                # submit a number of jobs less than the user's queue limit
+                util.kill_running_and_waiting_jobs(self.cook_url, user.name, log_jobs=False)
```
[the remark this answers is not in the document]
Comment (dposada) [reply integration/tests/cook/test_multi_user.py:994]: Done

File: integration/tests/cook/util.py [hunk 20]
```diff
@@ -1861,15 +1861,18 @@ def has_one_agent():
...
 
-def kill_running_and_waiting_jobs(cook_url, user):
+
+def kill_running_and_waiting_jobs(cook_url, user, log_jobs=True):
     one_hour_in_millis = 60 * 60 * 1000
     start = current_milli_time() - (72 * one_hour_in_millis)
     end = current_milli_time() + one_hour_in_millis
-    running = jobs(cook_url, user=user, state=['running', 'waiting'], start=start, end=end).json()
-    logger.info(f'Currently running/waiting jobs: {json.dumps(running, indent=2)}')
+    running = jobs(cook_url, user=user, state=['running', 'waiting'], start=start, end=end, limit=100000).json()
```
[the remark this answers is not in the document]
Comment (dposada) [reply integration/tests/cook/util.py]: Good call, made it try again

File: scheduler/src/cook/components.clj [hunk 21]
```diff
@@ -103,6 +103,11 @@
                           compute-clusters curator-framework mesos-datomic-mult leadership-atom
                           mesos-agent-attributes-cache pool-name->pending-jobs-atom mesos-heartbeat-chan
                           trigger-chans]
+
+                      ; We track queue limits on all nodes, not just the leader, because
+                      ; we need to check them when job submission requests come in
+                      ((util/lazy-load-var 'cook.queue-limit/start-updating-queue-lengths))
```
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/src/cook/components.clj]: Apparently nothing, un-lazified now

File: scheduler/src/cook/rest/api.clj [hunk 22]
```diff
@@ -1969,6 +1973,51 @@
...
+  this check does not take into account any of the following in real-time, and therefore is
+  only 100% accurate as of the last refresh from the database:
+
+  - Jobs will get removed from the queue when they launch
+  - Jobs will get added to the queue if they fail and have retries
+  - If there are multiple API hosts in the cluster, each one is maintaining a separate
+    count of the queue, without knowing about the other API hosts' job submissions or
+    job kills
+
+  Why is it OK to not take these things into account? This check is meant to prevent
```
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/src/cook/rest/api.clj]: Done

File: scheduler/src/cook/regexp_tools.clj (hunk 16 above)
Comment (dposada) [reply scheduler/src/cook/regexp_tools.clj:38]: When nothing matches, the regexp code doesn't throw, it returns `nil`:

```
(cook.regexp-tools/match-based-on-regexp :pool-regex :field-name [{:pool-regex "^pool-name$" :field-name "foo"}] "pool-name")
=> "foo"

(cook.regexp-tools/match-based-on-regexp :pool-regex :field-name [{:pool-regex "^pool-name$" :field-name "foo"}] "non-match")
=> nil
```

File: scheduler/test/cook/test/regexp_tools.clj (hunk 17 above)
Comment (dposada) [reply scheduler/test/cook/test/regexp_tools.clj]: Added unit tests for the defaulting code.

We have a pattern (prior to this PR) where `nil` is returned when no match is found, and I think that is appropriate and can remain.

File: scheduler/src/cook/regexp_tools.clj (hunk 16 above)
Comment (dposada) [reply scheduler/src/cook/regexp_tools.clj:38]: We have a pattern (prior to this PR) where `nil` is returned when no match is found, and I think that is appropriate and can remain.

Comment (dposada) [reply scheduler/src/cook/regexp_tools.clj:38]: I don't think we need to support distinguishing between a configured match where the configured value is `nil` vs. no match.

File: scheduler/test/cook/test/queue_limit.clj (hunk 15 above)
Comment (dposada) [reply scheduler/test/cook/test/queue_limit.clj]: Good call, done

File: scheduler/test/cook/test/queue_limit.clj [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+             (queue-limit/query-queue-lengths))))))
+
+(deftest test-inc-queue-length!
+  (testing "throws on nil pool name"
+    (is (thrown? AssertionError (queue-limit/inc-queue-length! nil "alice" 2)))))
+
+(deftest test-dec-queue-length!
+  (with-redefs [queue-limit/get-pending-jobs (constantly [])]
+    (queue-limit/update-queue-lengths!))
+  (queue-limit/inc-queue-length! "pool-1" "alice" 2)
```
[the remark this answers is not in the document]
Comment (dposada) [reply scheduler/test/cook/test/queue_limit.clj]: Split up with newlines and comments describing each section
